third_party/move/move-compiler-v2/src/pipeline/explicate_drop.rs

```diff
@@ -0,0 +1,190 @@
+        for arg in self.target.get_parameters() {
+            if !live_var_info.before.contains_key(&arg) && !lifetime_info.before.is_borrowed(arg) {
+                let attr_id = self.target.get_bytecode()[0].get_attr_id();
+                self.drop_temp(arg, attr_id)
+            }
+        }
+    }
+
+    // Returns a set of locals that can be dropped at given code offset
+    fn released_temps_at(&self, code_offset: CodeOffset) -> BTreeSet<TempIndex> {
+        let live_var_info = self.get_live_var_info(code_offset);
+        let lifetime_info = self.get_lifetime_info(code_offset);
+        let bytecode = &self.target.get_bytecode()[code_offset as usize];
+        released_temps(live_var_info, lifetime_info, bytecode)
+    }
+
+    fn get_live_var_info(&self, code_offset: CodeOffset) -> &'a LiveVarInfoAtCodeOffset {
+        self.live_var_annot
+            .get_live_var_info_at(code_offset)
+            .expect("live var info")
+    }
+
+    fn get_lifetime_info(&self, code_offset: CodeOffset) -> &'a LifetimeInfoAtCodeOffset {
+        self.lifetime_annot.get_info_at(code_offset)
+    }
+
+    fn drop_temp(&mut self, tmp: TempIndex, attr_id: AttrId) {
+        let drop_t = Bytecode::Call(attr_id, Vec::new(), Operation::Destroy, vec![tmp], None);
+        self.emit_bytecode(drop_t)
+    }
+
+    fn drop_temps(&mut self, temps_to_drop: &BTreeSet<TempIndex>, attr_id: AttrId) {
+        for t in temps_to_drop {
+            self.drop_temp(*t, attr_id)
+        }
+    }
+
+    fn emit_bytecode(&mut self, bytecode: Bytecode) {
+        self.transformed.push(bytecode)
+    }
+}
+
+// Returns a set of locals that can be dropped
+// these are the ones no longer alive or borrowed
+fn released_temps(
+    live_var_info: &LiveVarInfoAtCodeOffset,
+    life_time_info: &LifetimeInfoAtCodeOffset,
+    bytecode: &Bytecode,
+) -> BTreeSet<TempIndex> {
+    // use set to avoid duplicate dropping
+    let mut released_temps = BTreeSet::new();
+    for t in live_var_info.released_temps() {
+        if !life_time_info.after.is_borrowed(t) {
+            released_temps.insert(t);
+        }
+    }
+    for t in life_time_info.released_temps() {
+        if !live_var_info.after.contains_key(&t) {
+            released_temps.insert(t);
+        }
+    }
+    // if a temp is moved, then no need to drop
+    // this should come before the calculation
+    // of unused vars; because of, for instance,
+    // x = move(x)
+    for moved_src in moved_srcs(bytecode) {
+        released_temps.remove(&moved_src);
+    }
+    // this is needed because unused vars are not released by live var info
+    for dst in bytecode.dests() {
+        if !live_var_info.before.contains_key(&dst)
+            && !live_var_info.after.contains_key(&dst)
+            && !life_time_info.before.is_borrowed(dst)
+            && !life_time_info.after.is_borrowed(dst)
+        {
+            // todo: triggered in ability-checker/ability_violation.move
```

Comment:
Please use capital TODO for better searching

Reply:
done